pow/src/main/java/tech/pegasys/artemis/pow/Eth1DataManager.java

```diff
@@ -0,0 +1,372 @@
+
+                SafeFuture<UnsignedLong> newBlockNumberDiffFuture =
+                    getBlockNumberDiffWithMidRangeBlock(
+                        latestBlockTimestampFuture, realSecondsPerEth1BlockFuture, cacheMidRange);
+
+                middleBlockFuture =
+                    getMidRangeBlock(latestBlockNumberFuture, newBlockNumberDiffFuture);
+              }
+              return middleBlockFuture;
+            })
+        .thenCompose(
+            middleBlock -> {
+              EthBlock.Block block = middleBlock.getBlock();
+              UnsignedLong middleBlockNumber = UnsignedLong.valueOf(block.getNumber());
+              postCacheEth1BlockEvent(middleBlockNumber, block).reportExceptions();
+              SafeFuture<StartupLogicStates> exploreUpResultFuture =
+                  exploreBlocksInDirection(middleBlockNumber, true);
+              SafeFuture<StartupLogicStates> exploreDownResultFuture =
+                  exploreBlocksInDirection(middleBlockNumber, false);
+              return SafeFuture.allOf(exploreUpResultFuture, exploreDownResultFuture)
+                  .thenApply(
+                      done -> {
+                        StartupLogicStates exploreUpResult = exploreUpResultFuture.getNow(null);
+                        checkNotNull(exploreUpResult);
+
+                        StartupLogicStates exploreDownResult = exploreDownResultFuture.getNow(null);
+                        checkNotNull(exploreDownResult);
+
+                        if (exploreDownResult.equals(StartupLogicStates.DONE_EXPLORING)
+                            && exploreUpResult.equals(StartupLogicStates.DONE_EXPLORING)) {
+                          return StartupLogicStates.SUCCESSFULLY_COMPLETED;
+                        } else {
+                          return StartupLogicStates.UNSUCCESSFULLY_COMPLETED;
+                        }
+                      });
+            });
+  }
+
+  private SafeFuture<StartupLogicStates> exploreBlocksInDirection(
+      UnsignedLong blockNumber, final boolean isDirectionUp) {
+    blockNumber =
+        isDirectionUp ? blockNumber.plus(UnsignedLong.ONE) : blockNumber.minus(UnsignedLong.ONE);
+    SafeFuture<EthBlock> blockFuture = getEth1BlockFuture(blockNumber);
+    UnsignedLong finalBlockNumber = blockNumber;
+    return blockFuture
+        .thenCompose(
+            ethBlock -> {
+              EthBlock.Block block = ethBlock.getBlock();
+              if (isDirectionUp) latestBlockReference.set(block);
+              UnsignedLong timestamp = UnsignedLong.valueOf(block.getTimestamp());
+              postCacheEth1BlockEvent(finalBlockNumber, block).reportExceptions();
+              if (isTimestampInRange(timestamp)) {
+                return exploreBlocksInDirection(finalBlockNumber, isDirectionUp);
+              }
+              return SafeFuture.completedFuture(StartupLogicStates.DONE_EXPLORING);
+            })
+        .exceptionally(err -> StartupLogicStates.UNABLE_TO_EXPLORE_BLOCKS);
```

Comment:
Leave the exception to propagate back out as a failed future here rather than converting to a state and losing the detail of what exception was thrown.

Reply:
Done.